crates/hotfix/tests/common/mock_counterparty.rs

```diff
@@ -1,67 +1,74 @@
-use std::sync::Arc;
-use std::time::{Duration, Instant};
-use tokio::sync::mpsc::Receiver;
-use tokio::sync::{Mutex, mpsc, oneshot};
+use std::time::Duration;
+use tokio::sync::{mpsc, oneshot};
+
+const DEFAULT_TIMEOUT: Duration = Duration::from_secs(1);
 
 pub struct MockCounterparty {
-    received_messages: Arc<Mutex<Vec<RawFixMessage>>>,
+    // Receiver-End of the channel
+    receiver: mpsc::UnboundedReceiver<WriterMessage>,
+    // History of Received messages from the client
+    messages: Vec<WriterMessage>,
     _connection: FixConnection,
     _dc_sender: oneshot::Sender<()>,
 }
 
-type MessageStore = Arc<Mutex<Vec<RawFixMessage>>>;
-
 impl MockCounterparty {
     pub async fn start(session_ref: SessionRef<impl FixMessage>) -> Self {
-        let (writer_ref, received_messages) = Self::spawn_writer();
+        let (writer_ref, receiver) = Self::spawn_writer();
         let (reader_ref, dc_sender) = Self::create_reader();
         let connection = FixConnection::new(writer_ref, reader_ref);
 
         session_ref.register_writer(connection.get_writer()).await;
 
         Self {
-            received_messages,
+            receiver,
+            messages: vec![],
             _connection: connection,
             _dc_sender: dc_sender,
         }
     }
 
-    pub async fn assert_message_count(&self, expected_count: usize, timeout_secs: f32) {
-        let timeout_duration = Duration::from_secs_f32(timeout_secs);
-        let start_time = Instant::now();
-
-        loop {
-            {
-                let messages = self.received_messages.lock().await;
-                if messages.len() >= expected_count {
-                    return;
-                }
-            }
+    /// Listen to the next message on the channel
+    pub async fn get_next(&mut self, timeout: Option<Duration>) -> Option<&WriterMessage> {
+        let timeout = timeout.unwrap_or(DEFAULT_TIMEOUT);
+        let msg = tokio::time::timeout(timeout, self.receiver.recv())
+            .await
+            .unwrap_or_else(|_| panic!("Message not received in less than {timeout:?}"));
+        if let Some(msg) = msg {
+            self.messages.push(msg);
+            self.messages.last()
+        } else {
+            None
+        }
+    }
 
-            if start_time.elapsed() >= timeout_duration {
-                let current_count = self.received_messages.lock().await.len();
-                panic!(
-                    "Expected {expected_count} messages, but only received {current_count} within {timeout_secs} seconds"
-                );
-            }
+    pub async fn assert_next<F>(&mut self, timeout: Option<Duration>, assertion: F)
+    where
+        F: FnOnce(Option<&WriterMessage>) -> bool,
+    {
+        let msg = self.get_next(timeout).await;
+        assert!(assertion(msg));
+    }
 
-            tokio::time::sleep(Duration::from_millis(1)).await;
-        }
+    pub async fn assert_disconnected(&mut self) {
```

Comment:
This will probably also need a timeout, as you'll typically need to wait for the disconnect to happen after your last action in the test. But we can follow up on this when we actually have a use for this function.

Reply:
Your example in the test case contradicts me :) but that's essentially the case when the counterparty disconnects. We also have cases where the disconnect happens in response to, e.g. a logout flow.

I think in those cases you'll need a timeout, but I guess we'll see.